python/tskit/trees.py

```diff
@@ -1981,6 +1981,33 @@ def __next__(self):
         return self.tree
 
 
+class _ReversibleIterator(object):
+    """
+    Simple class to provide a reversible iterator over getters like ts.node() & ts.edge()
+    """
+    def __init__(self, getter, getter_length):
+        self.index = 0
+        self.getter = getter
+        self.getter_length = getter_length
+
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        return self.getter_length
+
+    def __getitem__(self, i):
+        return self.getter(i)
+
+    def __next__(self):
+        try:
+            element = self[self.index]
+        except IndexError:
```

Comment:
This depends on the getters always returning an index error, which I'm not sure we want to guarantee long term. It would be safer to check the bounds explicitly I think. What about negative indexes then? This stuff is quite subtle, we need to be 100% sure that we're implementing these protocols correctly.

Reply:
We should never get a negative self.index, as we are starting at 0 and incrementing upwards on each `next` loop.

Reply:
We don't need ``__next__`` at all, I think. Just implenent ``__len__`` and `__getitem__`` like [here](https://github.com/tskit-dev/tskit/blob/master/python/tskit/tables.py#L151) and delete all the other special methods. This should do everything we need.

Also need some tests of using negative indexes on the returned sequences then, of course.

Reply:
OK. Happy to do that, although it does make this a breaking change, if anyone was previously relying on the `next()` functionality